server/src/main/java/org/jetbrains/bsp/bazel/server/BazelBspServer.java

```diff
@@ -39,23 +42,63 @@ public class BazelBspServer {
 
   private final BazelRunner bazelRunner;
   private final BazelInfo bazelInfo;
-
   private final BspServerApi bspServerApi;
   private final BazelBspCompilationManager compilationManager;
   private final BspClientLogger bspClientLogger;
 
-  public BazelBspServer(WorkspaceContextProvider workspaceContextProvider) {
-    this.bspClientLogger = new BspClientLogger();
-    var bspInfo = new BspInfo();
-    var bazelInfoStorage = new BazelInfoStorage(bspInfo);
-    var bazelDataResolver =
-        new BazelInfoResolver(
-            BazelRunner.inCwd(workspaceContextProvider, bspClientLogger), bazelInfoStorage);
-    this.bazelInfo = bazelDataResolver.resolveBazelInfo();
-    this.bazelRunner = BazelRunner.of(workspaceContextProvider, bspClientLogger, bazelInfo);
+  public BazelBspServer(
+      BspInfo bspInfo,
+      WorkspaceContextProvider workspaceContextProvider,
+      Option<Path> workspaceRoot) {
+    var serverContainer =
+        makeProjectProvider(bspInfo, workspaceContextProvider, workspaceRoot, Option.none());
+    this.bspClientLogger = serverContainer.bspClientLogger;
+    this.bazelInfo = serverContainer.bazelInfo;
+    this.compilationManager = serverContainer.compilationManager;
+    this.bazelRunner = serverContainer.bazelRunner;
+
+    var bspProjectMapper = new BspProjectMapper(serverContainer.languagePluginsService);
+    var projectSyncService =
+        new ProjectSyncService(bspProjectMapper, serverContainer.projectProvider);
+    var executeService =
+        new ExecuteService(compilationManager, serverContainer.projectProvider, bazelRunner);
+    var cppBuildServerService = new CppBuildServerService();
     var serverLifetime = new BazelBspServerLifetime();
     var bspRequestsRunner = new BspRequestsRunner(serverLifetime);
-    this.compilationManager = new BazelBspCompilationManager(bazelRunner);
+
+    this.bspServerApi =
+        new BspServerApi(
+            serverLifetime,
+            bspRequestsRunner,
+            projectSyncService,
+            executeService,
+            cppBuildServerService);
+  }
+
+  public static ServerContainer makeProjectProvider(
+      BspInfo bspInfo,
+      WorkspaceContextProvider workspaceContextProvider,
+      Option<Path> workspaceRoot,
+      Option<ProjectStorage> projectStorage) {
+    var bspClientLogger = new BspClientLogger();
+    var bazelInfoStorage = new BazelInfoStorage(bspInfo);
+    var bazelDataResolver =
+        workspaceRoot
```

Comment:
Not sure what happens here. The idea of bazel runner constructors was that `inCwd` just runs bazel to resolve bazel info (and it is assumed that it is run somewhere inside bazel workspace), so it is used at most once, to construct proper bazel runner with workspace path that was extracted from bazel. IF now we have a better way to find workspace we can drop this behavior.